src/Parser.php

```diff
@@ -48,76 +46,100 @@ class Parser
 
 					# Others start #
 						(?>T # Literal T
-						(?<hourNum>2[0-3]|[01][0-9]):
-						(?<minuteNum>[0-5][0-9]):
-						(?<secondNum>[0-5][0-9])
+						(?<hour>2[0-3]|[01][0-9]):
+						(?<minute>[0-5][0-9]):
+						(?<second>[0-5][0-9])
 						(?>(?<tzUtc>Z)|
 						(?<tzSign>[+-])
 						(?<tzHour>[01][0-9]):
 						(?<tzMinute>[0-5][0-9]))?)?)?)?$
 					# Others end #
 					/";
 
-    public function parse(string $data): DateTimeInterface
+
+    private ?int $year = null;
+    private ?int $month = null;
+    private ?int $day = null;
+    private ?int $hour = null;
+    private ?int $minute = null;
+    private ?int $second = null;
+    private ?string $tzSign = null;
+    private ?int $tzMinute = null;
+    private ?int $tzHour = null;
+    private ?string $tzUtc = null;
```

Comment:
Adding these fields and corresponding getters here has me just as confused as requiring a Parser instance in the date obejcts. Why end up with such state in the Parser?

Reply:
I want a ```Parser``` just to do the parsing and store the results. This results should be then used by EDTF class to generate an ```ExtDate```, ```ExtDateTime```, or an ```Interval``` object

Reply:
I think it would be better if this state was put in a dedicated object. That way the parser could hand that object to the EDTF class and remain state free itself. Having such state in a service like Parser is a design anti-pattern IMO. It opens the door for a whole class of state bugs and makes construction of the EDTF objects unnecessarily difficult.

At the same time I'd be weird for the Parser to return an intermediate object. Why can't it return the real object model (`ExtDate`, `ExtDateTime`, `Interval`) right away?

Reply:
Actually the old Parser returns a real object (```ExtDate```, ```ExtDateTime```, ```Interval```) like you mention above:

https://github.com/ProfessionalWiki/EDTF/blob/d3c0268be34466876480a04c6db1bb014acec7c3/src/Parser.php#L61-L67

I have change it, because you suggest ```Parser::parse()``` to returns a ```ParserResult``` object like in this pull:
https://github.com/ProfessionalWiki/EDTF/pull/4
Instead creating new Class I am thinking why not just having this Parser class to store all of the parsing result values. That's why this Parser class have properties like that.

Reply:
Normally classes should only do one thing. Or have a **S**ingle Responsibility, the S in SOLID. Don't be afraid to create many classes or small classes, this is nearly always better than mashing things together or otherwise going against SOLID and similar standard software design practices. (Most devs don't know "standard" about design principles so the "standard" perhaps needs some quotes.)

I think the Parser should return an EDTF date, which is a Value Object. Possibly the date can be wrapped in a result object like the one I created an example off. The Parser should not expose internal state or require the caller to use another service such as a factory to turn the return value into an EDTF date. The parser _can_ be split up into multiple classes internally (if this makes things easier), but this should not be visible to the caller of the Parser (including most of the tests).

I have been wondering for some time what the `EDTF` class is all about. It does not seem needed. On your branch at the moment my impression is that it violates Single Responsibility by splitting up a responsibility over multiple classes. Classes full of static code are also very suspect. OO is about objects, not classes :)

Reply:
So your decision is to let Parser act as factory like in this code?
https://github.com/ProfessionalWiki/EDTF/blob/d3c0268be34466876480a04c6db1bb014acec7c3/src/Parser.php#L61-L67

Reply:
Yes

Reply:
Done in https://github.com/ProfessionalWiki/EDTF/pull/5/commits/145d95a4d9092dbd9e10af2ff4d8fe8077eb33c7